semgrep-core/src/core/Rule.ml

```diff
@@ -226,20 +216,34 @@ and paths = {
 }
 [@@deriving show]
 
+type search_rule = [ `Search of common * pformula ] [@@deriving show]
+type taint_rule = [ `Taint of common * taint_spec ] [@@deriving show]
+type rule = [ search_rule | taint_rule ] [@@deriving show]
+
 (* alias *)
 type t = rule [@@deriving show]
 type rules = rule list [@@deriving show]
 
+let id (`Search (c, _) | `Taint (c, _)) = c.id
+let message (`Search (c, _) | `Taint (c, _)) = c.message
+let severity (`Search (c, _) | `Taint (c, _)) = c.severity
+let languages (`Search (c, _) | `Taint (c, _)) = c.languages
+let options (`Search (c, _) | `Taint (c, _)) = c.options
+let fix (`Search (c, _) | `Taint (c, _)) = c.fix
+let fix_regexp (`Search (c, _) | `Taint (c, _)) = c.fix_regexp
+let paths (`Search (c, _) | `Taint (c, _)) = c.paths
+let metadata (`Search (c, _) | `Taint (c, _)) = c.metadata
+
 (*****************************************************************************)
 (* Helpers *)
 (*****************************************************************************)
 
-let partition_rules rules =
+let partition_rules (rules : rules) =
```

Comment:
Don't need to fix it in this PR but this function seems kind of pointless now, since it basically just renames constructors

Reply:
I think it's still useful, because it lets us go from a mixed list to two unmixed lists (N-many for N modes).

Reply:
actually just kidding I did not know what `partition` does lol